README.md
# ScopeDog eFinder

- [ScopeDog eFinder](#scopedog-efinder)
  - [Needed parts](#needed-parts)
    - [Purchase of the hardware](#purchase-of-the-hardware)
  - [Nexus DSC Pro](#nexus-dsc-pro)
    - [Update of the Nexus DSC Pro firmware](#update-of-the-nexus-dsc-pro-firmware)
    - [Network settings on the Nexus DSC Pro](#network-settings-on-the-nexus-dsc-pro)
  - [Software](#software)
    - [Raspberry Pi 4](#raspberry-pi-4)
      - [Install OS and needed dependencies](#install-os-and-needed-dependencies)
      - [Install the ZWO ASI Linux SDK](#install-the-zwo-asi-linux-sdk)
      - [Start eFinder](#start-efinder)
      - [Start eFinder automatically after boot](#start-efinder-automatically-after-boot)
      - [Install RTL8192EU driver for the TP-LINK TL-WN823N](#install-rtl8192eu-driver-for-the-tp-link-tl-wn823n)
    - [The Raspberry Pi Pico handbox](#the-raspberry-pi-pico-handbox)
    - [Changelog](#changelog)
  - [Hardware](#hardware)
    - [Adding the OLED display to the Raspberry Pi Pico](#adding-the-oled-display-to-the-raspberry-pi-pico)
    - [Connect to the Nexus DSC wifi](#connect-to-the-nexus-dsc-wifi)
    - [How to work with the handpad](#how-to-work-with-the-handpad)
      - [In the VNC GUI version](#in-the-vnc-gui-version)
      - [In the handpad version](#in-the-handpad-version)
  - [Development environment](#development-environment)
    - [Poetry](#poetry)
    - [Docker](#docker)

## Needed parts

### Purchase of the hardware

As I live in Belgium, this section is very Europe-centric.  I ordered the ASI camera at Teleskop Service (Germany) and the rest in the Netherlands.

| Ordered    | Product                                                                                                                                                                                    | Company           | Amount   | Arrival   |
| ---------- | ------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------ | ----------------- | -------- | ---------- |
| 03/08/2022 | Software                                                                                                                                                                                   | Keith Venables    | 26.98 €  | 03/08/2022 |
| 03/08/2022 | [ASI120MM-S](https://www.teleskop-express.de/shop/product_info.php/info/p7109)                                                                                                             | Teleskop Service  | 248.51 € | 11/08/2022           |
| 03/08/2022 | [Raspberry Pi 4 model B - 4GB](https://www.raspberrystore.nl/PrestaShop/nl/raspberry-pi-v4/228-raspberry-pi-4b4gb-765756931182.html?search_query=Raspberry+Pi+4+model+B+-+4GB&results=203) | Raspberrystore.nl | 74.95 €  | 10/08/2022           |
| 03/08/2022 | [Raspberry Pi Pico](https://www.raspberrystore.nl/PrestaShop/nl/raspberry-pi-pico/312-raspberry-pi-pico-0617588405587.html?search_query=Raspberry+Pi+4+Pico&results=203)                   | Raspberrystore.nl | 3.99 €   | 10/08/2022           |
| 03/08/2022 | [2 x 1x40 header](https://www.raspberrystore.nl/PrestaShop/nl/raspberry-pi-pico/320-soldeerpennen-1x40-voor-de-raspberry-pi-pico.html?search_query=1x40+header&results=44)                 | Raspberrystore.nl | 2.00 €   | 10/08/2022           |
| 03/08/2022 | Verzendkosten Astroberrystore.nl                                                                                                                                                           | Raspberrystore.nl | 12.74 €  | 10/08/2022           |
| 03/08/2022 | [2.23 OLED Display Module for Raspberry Pi Pico](https://www.amazon.nl/gp/product/B093SYSX5S/ref=ppx_od_dt_b_asin_title_s00?ie=UTF8&psc=1)                                                 | Amazon (.nl)      | 20.99 €  | 09/08/2022          |
| 03/08/2022 | [Geekworm Raspberry Pi 4 Aluminum Case](https://www.amazon.nl/gp/product/B07ZVJDRF3/ref=ppx_od_dt_b_asin_title_s01?ie=UTF8&psc=1)                                                          | Amazon (.nl)      | 14.89 €  | 08/08/2022           |
| 03/08/2022 | [Sandisk Ultra 32 GB microSDHC](https://www.amazon.nl/gp/product/B08GY9NYRM/ref=ppx_od_dt_b_asin_title_s02?ie=UTF8&psc=1)                                                                  | Amazon (.nl)      | 8.01 €   | 08/08/2022           |
| 03/08/2022 | [100 x 60 x 25 mm DIY box](https://www.amazon.nl/gp/product/B07V2Q32H8/ref=ppx_od_dt_b_asin_title_s02?ie=UTF8&psc=1)                                                                       | Amazon (.nl)      | 10.62 €  | 08/08/2022           |
| 03/08/2022 | [Lon0167  Momentary Circuit Control Tactile Tact Push Button Switch](https://www.amazon.nl/gp/product/B0842JYXC8/ref=ppx_yo_dt_b_asin_title_o00_s00?ie=UTF8&psc=1)                         | Amazon (.nl)      | 11.31 €  | 16/08/2022           |
| 03/08/2022 | [50 mm F1.8 CCTV-lens for C-mount](https://www.amazon.nl/gp/product/B08BF7DRXR/ref=ppx_yo_dt_b_asin_title_o02_s00?ie=UTF8&psc=1)                                                           | Amazon (.nl)      | 48.38 €  | 16/08/2022           |
| 10/10/2022 | [TP-Link 300 Mbps Wifi USB-adapter (TL-WN823N)](https://www.amazon.nl/dp/B0088TKTY2/ref=pe_28126711_487805961_TE_item)                                                                     | Amazon (.nl)      | 9.95 €  | 11/10/2022           |
|Total      |                                                                                                                                                                                            |                   | 493.32 € |            |

## Nexus DSC Pro

### Update of the Nexus DSC Pro firmware

- Remove the micro SD card from the Nexus DSC Pro.
- Insert the micro SD card into a card reader of your computer.
- Make sure that at least version 1.1.18 of the firmware is installed on your Nexus DSC Pro.  If this is not the case, download the latest version of the firmware from the Nexus DSC pro webpage and copy the firmware image (nxpro.fw) to the micro SD card.
- Make sure you properly eject the micro SD card from your computer.
- Insert the micro SD card back into the slot of the Nexus DSC Pro.
- Press and hold the OK button and turn the Nexus DSC Pro.
- A screen is shown asking if it is OK to update the firmware.  Press OK and wait till the update is completed.
- In the info menu, the correct version of the firmware should be shown (1.1.18).
- In version 1.1.18 of the Nexus DSC Pro firmware, there is a problem with the alignment. An easy workaround is to select 'two stars w/o Alt Ref’ in Align set up.

### Network settings on the Nexus DSC Pro

Make sure the following settings are set on the Network page:

- WiFi Mode: Access Point
- DHCP: Disabled
- SSID: Name the network
- Password: Select a password or leave blank
- Channel: 6 (The channel needs to be 6 to be able to connect the eFinder, a VNC client, and SkySafari on a tablet)
- Protocol: LX200 (ServoCAT will not connect with the eFinder)

## Software

### Raspberry Pi 4

#### Install OS and needed dependencies

- Download the [Raspberry Pi Imager](https://www.raspberrypi.com/software/)
- Select the standard 32bit Raspberry Pi OS.
- In the settings, enter *efinder* as username, *efinder* as password, and *efinder.local* as hostname.
- Start the installation of the OS.
- Checkout the eFinder software from GitHub

```bash
cd ~
git clone https://github.com/WimDeMeester/eFinder.git eFinder
```

- Install all dependencies

```bash
cd eFinder
sudo ./pi-install.sh
```

- Change the resolution for the vnc server, by starting

```bash
sudo raspi-config
```

- Reboot

#### Install the ZWO ASI Linux SDK

- Download the SDK from the Developer tab of the [driver page](https://astronomy-imaging-camera.com/software-drivers).
- Unpack the SDK:

```bash
wget "https://dl.zwoastro.com/software?app=AsiCameraDriverSdk&platform=macIntel&region=Overseas" -O ASI_linux_mac_SDK.tar.bz2
bunzip2 ASI_linux_mac_SDK.tar.bz2
tar xvf ASI_linux_mac_SDK.tar
cd ASI_linux_mac_SDK_V1.27/lib/
sudo mkdir /lib/zwoasi
sudo cp -r * /lib/zwoasi/
sudo install asi.rules /lib/udev/rules.d
```

#### Start eFinder

Activate a python virtual environment (do this every time):

`poetry shell`

Run the VNC Gui version of the app:

`python src/eFinderVNCGUI.py`

The headless version can be started using the command:

`python src/eFinder.py`

Run the VNC Gui version of the app without having a handpad, a camera or a nexus device.
(The headless version has no command line options at the moment)

`python src/eFinderVNCGUI.py -fh -fn -fc`

An easier way to start the app is executing **scripts/start.sh** for the headless version and **scripts/startVNCGUI.sh** for the GUI version.

#### Start eFinder automatically after boot

- Find the PATH

```bash
echo $PATH
```

- Adapt crontab:

```bash
crontab -e
```

- Add the PATH and the DISPLAY variable to the crontab

```bash
export PATH=<The returned path from the echo $PATH command>
DISPLAY=:0
@reboot sleep 20 && (/home/efinder/eFinder/scripts/start.sh > /home/efinder/logs.txt 2>&1)
```

It is not possible to start the VNC GUI version automatically from crontab.

#### Install RTL8192EU driver for the TP-LINK TL-WN823N

- Install the needed packages

```bash
sudo apt-get install git raspberrypi-kernel-headers build-essential dkms
```

- Clone the driver from the GitHub repository

```bash
git clone https://github.com/Mange/rtl8192eu-linux-driver
cd rtl8192eu-linux-driver
```

- Make sure the following lines are in the ***Makefile***:

```bash
CONFIG_PLATFORM_I386_PC = n
CONFIG_PLATFORM_ARM_RPI = y
```

- Add and install the driver to DKMS

```bash
sudo dkms add .
sudo dkms install rtl8192eu/1.0
```

- Make sure the driver is loaded correctly

```bash
echo "blacklist rtl8xxxu" | sudo tee /etc/modprobe.d/rtl8xxxu.conf
echo -e "8192eu\n\nloop" | sudo tee /etc/modules
```

- Fix possible plugging/replugging issue and sforce the driver to be active from boot:

```bash
echo -e "8192eu\n\nloop" | sudo tee /etc/modules
echo "options 8192eu rtw_power_mgnt=0 rtw_enusbss=0" | sudo tee /etc/modprobe.d/8192eu.conf;
```

- Reboot the system

```bash
sudo reboot
```

- Configure the new Wifi network
- Disable the Wifi card from the Raspberry Pi by adding the following lines to ***/etc/modprobe.d/raspi-blacklist.conf***

```bash
sudo bash -c 'echo "blacklist brcmfmac" >> /etc/modprobe.d/raspi-blacklist.conf'
sudo bash -c 'echo "blacklist brcmutil" >> /etc/modprobe.d/raspi-blacklist.conf'
```

- Reboot the system

```bash
sudo reboot
```

### The Raspberry Pi Pico handbox

The software for the Raspberry Pi Pico handbox is written in micropython.  We first need to install micropython on the board.

- Download *Thonny* from the [website](https://thonny.org/).
- Connect the Raspberry Pi Pico to your computer, while pushing the reset button.
- Start *Thonny* and select to install micropython (at the bottom right).
- The Raspberry Pi Pico will automatically reboot and install micropython.
- After the Raspberry Pi Pico was restarted, open **main.py** in Thonny and select **Save As...**.  Select the Raspberry Pi Pico and save the file as **main.py**.  The screen of the handbox should show *ScopeDog eFinder - No eFinder yet*.

### Changelog

The changelog of the software can be found [here](CHANGELOG.md).

## Hardware

### Adding the OLED display to the Raspberry Pi Pico

- Solder the header to the Raspberry Pi Pico.  Check if you are soldering the header to the correct side of the Raspberry Pi Pico (I soldered it the wrong side first).
- Make sure to also solder a cable to GP16, GP17, GND, GP18, GP19, and GP21.
- Push the OLED display on the soldered header.

### Connect to the Nexus DSC wifi

- Set up the Pi wifi to ‘know’ the Nexus DSC wifi DSC & password. That way on Pi boot it will always look for the Nexus wifi and connect as a priority.

### How to work with the handpad

#### In the VNC GUI version

- Simple press: Capture/solve & display RA, Dec & deltas.
- Up: Nexus align or Local Sync
- Down: Goto++

#### In the handpad version

![image](doc/menu.png)

## Development environment

### Poetry

*[Install poetry](https://python-poetry.org/docs/#installation) to manage the python dependencies.*

Install the dependencies and the virtual environment (do this once):

`poetry install`

Activate a python virtual environment (do this every time):

`poetry shell`

Run the VNC Gui version of the app without having a handpad, a camera or a nexus device.
(The headless version has no command line options at the moment)

`python eFinderVNCGUI.py -fh -fn -fc`

### Docker

Building the image:
```docker build --no-cache --platform linux/arm64/v8  -t efinder .```

Running the image:
```docker run -d --name efinder-test efinder```

Accessing the running image:
```docker exec -it efinder-test bash```


CHANGELOG.md
# Changelog for eFinder software

## Version 17.0

[Mike Rosseel]

- Extracted common functions from eFinder.py and eFinderVNCGui.py
- Extracted platesolving to it's own class and used that class in eFinder.py and eFinderVNCGui.py
- removed all Solver dir references, files are now either in the eFinder directory, in the Stills folder or on /dev/shm/images
- used Path everywhere for cleaner file handling
- Introduced argparse to eFinderVNCGui.py to make CLI configuration easier, and later down the road merge it with eFinder.py
- Added Debug classes for Camera, Handpad and Nexus, so that the program can be started without these attached.
- added poetry dependency management

[Keith Venables]

- Issue #27: eFinder did not handle negative declination values
- Issue #31: Bug in check align status
- Issue #32: More rugged load of parameters

[Wim De Meester]

- Issue #36: Aligning stars does not work...
- Issue #30: Startup scripts for eFinder.py

## Version 16_3

[Keith Venables]

- Resolved problem with Nexus DSC not returning its own GoTo target coordinates (using LX200 protocol). New Nexus firmware released on 17th Nov 2022
- Cleaned up GUI panel, removing redundant check boxes (finder focallength and goto source). Moved graticule checkbox to same sub-panel other display options.
- Added new test images (test.jpg & polaris.jpg) that match the 50mm default focal length. Removed all code references to choice of 200mm lens.
- Handpad now offers basic functionality even when the GUI is being run. (Solve, Align & Goto++)
- removed bug whereby in test mode the wrong scale was being used to generate offsets.

## Version 16

[Keith Venables]

- Added ability for user to adjust OLED brightness on the handpad. (main.py)
- Fixed a bug where the RA & Dec wasnt being tagged to the saved image file name.
	Nexus.py

## Version 15

[Keith Venables]

- Added support for QHY cameras (only tested so far with a QHY5L-II Mono)
	install the following to ~/Solver
	libqhy.py
	QHYCamera.py
	qhyccd.py
	add extra line to efinder.config.... Camera Type ('QHY' or 'ASI'):ASI
- Added option to choose between Nexus & SkySafari for source of GoTo target.
	add extra line to efinder.config.... SkySafari GoTo++:0
	0 = Nexus, 1 = SkySafari.
- removes option for 200mm focal length finder scope (ie 50mm only)
- removes support for LCD display module (ie OLED only)

## Version 14

[Keith Venables]

- Fixed a bug on the handpad version whereby in the final 'status' screen, use of the up or down buttons crashed the code.

- Fixed a bug that prevents handpad saved exposure & gain preferences from being displayed on the GUI.

- Modified how the handpad display starts when using the GUI version.

- Fixed a bug on the handpad version whereby a failed solve could result in the previous delta being displayed without a 'failed solve' warning

[Keith Venables]

- Make the handpad work with the VNC GUI version

[Wim De Meester]

- Move the Nexus code to a class outside of the main classes
- Create a Coordinates class to transform RA and dec to AltAz
- Create a handpad class with all code for the display box
- Create CameraInterface and ASICamera class

## Version 13

[Keith Venables, Wim De Meester]

The clock section in the GUI variant is now a thread and runs nice and smooth. (Wim’s code - Thx)

I’ve pulled out all the Nexus communications into a Class method. On boot the code also determines how it can connect to the Nexus, usb or wifi. Thus these variants aren't needed any more. This will make code changes a lot easier.

For instance the Nexus RA can be obtained by single call
ra = Nexus.get(:GR#)

Some Nexus commands don’t produce a reply, so they are

Nexus.write(:P#)

etc

Nexus.read() produces a grab of all the location and time data from the Nexus.

## Version 12

[Keith Venables]

Biggest recent issue has been the realisation that Skyfield wasn't precessing JNow back to J2000, despite it being described and no errors being thrown. This was messing up the align and local sync functions (adding quite a few arc minutes of error). Also could see that various methods of converting from RA & Dec to AltAz were producing different results.

The fix has been a significant change. The eFinder uses JNow exclusively now internally (solver output being precessed immediately) and my own altAz conversion method applied to all cases for consistency.

The rather neat method of finishing a classic goto, by doing a solve and local sync, followed by repeat goto (aka 'goto++'), wasnt consistent. Now realise that the above issues were partly to blame, plus a discovery that the LX200 protocol used to talk to the Nexus has one “target” used for both align and goto. Easily fixed by reading the goto target before the local sync, and then resending it back to the Nexus as part of the goto++.

This goto++ method seems now to be very consistent. The solve & local sync is very accurate (although Serge has a bug in the Nexus DSC for repeated syncs at the same location) and the goto++ results in scope positioning as good as your drive (ServoCat or ScopeDog) can manage. I’m getting about an arc minute.

The Nexus bug affects everybody with a  Nexus DSC - but not many will have noticed! Repeated local syncs result in the Nexus not quite resetting to the required RA & Dec. Can be 1/10 to 1/3 of a degree off. Without the eFinder most users wouldn’t be aware of the error - except the target wouldn’t be centred in the eyepiece. First local sync in any area of the sky is perfect. Serge is on the case.

For a while I have been using the wcs- modules available with the solver to make the offset measurement and application easier. (Eg the solver can return the RA & Dec of any pixel in the image, not just the centre.

I have now modified the initial offset calibration routine. Just align the main scope with any bright, named star, and the eFinder will recognise it and do the calibration. The offset measured can be saved to disk.


eFinder.py
#!/usr/bin/python3

# Program to implement an eFinder (electronic finder) on motorised Alt Az telescopes
# Copyright (C) 2022 Keith Venables.
# This program is free software: you can redistribute it and/or modify
# it under the terms of the GNU General Public License as published by
# the Free Software Foundation, either version 3 of the License, or
# (at your option) any later version.
# This program is distributed in the hope that it will be useful,
# but WITHOUT ANY WARRANTY; without even the implied warranty of
# MERCHANTABILITY or FITNESS FOR A PARTICULAR PURPOSE.  See the
# GNU General Public License for more details.
# This variant is customised for ZWO ASI ccds as camera, Nexus DSC as telescope interface
# It requires astrometry.net installed

import subprocess
import time
import os
import math
from PIL import Image
import psutil
import re
from skyfield.api import Star
import numpy as np
import threading
import select
from pathlib import Path
import fitsio
import Nexus
import Coordinates
import Display
import ASICamera
from platesolve import PlateSolve
from common import Common
import utils

cwd_path: Path = Path.cwd() 
images_path: Path = Path("/dev/shm/images")
utils.create_path(images_path) # create dir if it doesn't yet exist
# os.system('pkill -9 -f eFinder.py') # stops the autostart eFinder program running
x = y = 0  # x, y  define what page the display is showing
deltaAz = deltaAlt = 0
increment = [0, 1, 5, 1, 1]
offset_flag = False
align_count = 0
offset = 640, 480
star_name = "no star"
solve = False
sync_count = 0
pix_scale = 15
platesolve = PlateSolve(pix_scale, images_path)
common = Common(cwd_path=cwd_path, images_path=images_path, pix_scale=pix_scale, version_suffix="")
version = common.get_version()

def imgDisplay():  # displays the captured image on the Pi desktop.
    for proc in psutil.process_iter():
        if proc.name() == "display":
            proc.kill()  # delete any previous image display
    im = Image.open(images_path / "capture.jpg")
    im.show()


def solveImage():
    global offset_flag, solve, solvedPos, elapsed_time, star_name, star_name_offset, solved_radec, solved_altaz
    handpad.display("Started solving", "", "")
    result, elapsed_time = platesolve.solve_image(offset_flag)
    result = str(result.stdout)
    if "solved" not in result:
        print("Bad Luck - Solve Failed")
        handpad.display("Not Solved", "", "")
        solve = False
        return
    if (offset_flag == True) and ("The star" in result):
        table, h = fitsio.read(cwd_path / "capture.axy", header=True)
        star_name_offset = table[0][0], table[0][1]
        lines = result.split("\n")
        for line in lines:
            if line.startswith("  The star "):
                star_name = line.split(" ")[4]
                print("Solve-field Plot found: ", star_name)
                break
    solvedPos = common.applyOffset(nexus, offset)
    ra, dec, d = solvedPos.apparent().radec(coordinates.get_ts().now())
    solved_radec = ra.hours, dec.degrees
    solved_altaz = coordinates.conv_altaz(nexus, *(solved_radec))
    nexus.set_scope_alt(solved_altaz[0] * math.pi / 180)
    arr[0, 2][0] = "Sol: RA " + coordinates.hh2dms(solved_radec[0])
    arr[0, 2][1] = "   Dec " + coordinates.dd2dms(solved_radec[1])
    arr[0, 2][2] = "time: " + str(elapsed_time)[0:4] + " s"
    solve = True
    deltaCalc()


def deltaCalc():
    global deltaAz, deltaAlt, solved_altaz, scopeAlt, elapsed_time
    deltaAz, deltaAlt = common.deltaCalc(nexus.get_altAz(), solved_altaz, nexus.get_scope_alt(), deltaAz, deltaAlt)
    deltaXstr = "{: .2f}".format(float(deltaAz))
    deltaYstr = "{: .2f}".format(float(deltaAlt))
    arr[0, 3][0] = "Delta: x= " + deltaXstr
    arr[0, 3][1] = "       y= " + deltaYstr
    arr[0, 3][2] = "time: " + str(elapsed_time)[0:4] + " s"


def align():
    global align_count, solve, sync_count, param, offset_flag, arr
    new_arr = nexus.read_altAz(arr)
    arr = new_arr
    capture()
    imgDisplay()
    solveImage()
    if solve == False:
        handpad.display(arr[x, y][0], "Solved Failed", arr[x, y][2])
        return
    align_ra = ":Sr" + coordinates.dd2dms((solved_radec)[0]) + "#"
    align_dec = ":Sd" + coordinates.dd2aligndms((solved_radec)[1]) + "#"
    valid = nexus.get(align_ra)
    print(align_ra)
    if valid == "0":
        print("invalid position")
        handpad.display(arr[x, y][0], "Invalid position", arr[x, y][2])
        return
    valid = nexus.get(align_dec)
    print(align_dec)
    if valid == "0":
        print("invalid position")
        handpad.display(arr[x, y][0], "Invalid position", arr[x, y][2])
        return
    reply = nexus.get(":CM#")
    print("reply: ", reply)
    p = nexus.get(":GW#")
    print("Align status reply ", p)
    align_count += 1
    if p != "AT2#":
        handpad.display(
            "'select' aligns",
            "align count: " + str(align_count),
            "Nexus reply: " + p[0:3],
        )
    else:
        if p == "AT2#":
            sync_count += 1
            handpad.display(
                "'select' syncs",
                "Sync count " + str(align_count),
                "Nexus reply " + p[0:3],
            )
            nexus.set_aligned(True)
    return


def measure_offset():
    global offset_str, offset_flag, param, scope_x, scope_y, star_name
    offset_flag = True
    handpad.display("started capture", "", "")
    capture()
    imgDisplay()
    solveImage()
    if solve == False:
        handpad.display("solve failed", "", "")
        return
    scope_x = star_name_offset[0]
    scope_y = star_name_offset[1]
    d_x, d_y, dxstr, dystr = common.pixel2dxdy(scope_x, scope_y)
    param["d_x"] = d_x
    param["d_y"] = d_y
    save_param()
    offset_str = dxstr + "," + dystr
    arr[0, 5][1] = "new " + offset_str
    arr[0, 6][1] = "new " + offset_str
    handpad.display(arr[0, 5][0], arr[0, 5][1], star_name + " found")
    offset_flag = False


def up_down(v):
    global x
    x = x + v
    handpad.display(arr[x, y][0], arr[x, y][1], arr[x, y][2])


def left_right(v):
    global y
    y = y + v
    handpad.display(arr[x, y][0], arr[x, y][1], arr[x, y][2])


def up_down_inc(i, sign):
    global increment
    arr[x, y][1] = int(float(arr[x, y][1])) + increment[i] * sign
    param[arr[x, y][0]] = float(arr[x, y][1])
    handpad.display(arr[x, y][0], arr[x, y][1], arr[x, y][2])
    update_summary()
    time.sleep(0.1)


def flip():
    global param
    arr[x, y][1] = 1 - int(float(arr[x, y][1]))
    param[arr[x, y][0]] = str((arr[x, y][1]))
    handpad.display(arr[x, y][0], arr[x, y][1], arr[x, y][2])
    update_summary()
    time.sleep(0.1)


def update_summary():
    global param
    arr[1, 0][0] = "Ex:" + str(param["Exposure"]) + "  Gn:" + str(param["Gain"])
    arr[1, 0][1] = "Test mode:" + str(param["Test mode"])
    save_param()


def capture():
    global param
    extras = {}
    if param["Test mode"] == "1":
        if offset_flag == False:
            extras['testimage'] = 'm13'
        else:
            extras['testimage'] = 'polaris'
    radec = nexus.get_short()
    camera.capture(
        int(float(param["Exposure"]) * 1000000),
        int(float(param["Gain"])),
        radec,
        extras
    )


def go_solve():
    global x, y, solve, arr
    new_arr = nexus.read_altAz(arr)
    arr = new_arr
    handpad.display("Image capture", "", "")
    capture()
    imgDisplay()
    handpad.display("Plate solving", "", "")
    solveImage()
    if solve == True:
        handpad.display("Solved", "", "")
    else:
        handpad.display("Not Solved", "", "")
        return
    x = 0
    y = 3
    handpad.display(arr[0, 3][0], arr[0, 3][1], arr[0, 3][2])


def goto():
    handpad.display("Attempting", "GoTo++", "")
    goto_ra = nexus.get(":Gr#")
    if (
        goto_ra[0:2] == "00" and goto_ra[3:5] == "00"
    ):  # not a valid goto target set yet.
        print("no GoTo target")
        return
    goto_dec = nexus.get(":Gd#")
    print("Target goto RA & Dec", goto_ra, goto_dec)
    align()
    if solve == False:
        handpad.display("problem", "solving", "")
        return
    nexus.write(":Sr" + goto_ra + "#")
    nexus.write(":Sd" + goto_dec + "#")
    reply = nexus.get(":MS#")
    handpad.display("Performing", " GoTo++", "")
    time.sleep(5)  # replace with a check on goto progress
    go_solve()


def reset_offset():
    global param, arr
    param["d_x"] = 0
    param["d_y"] = 0
    offset_str = "0,0"
    arr[0, 5][1] = "new " + offset_str
    arr[0, 6][1] = "new " + offset_str
    handpad.display(arr[x, y][0], arr[x, y][1], arr[x, y][2])
    save_param()


def get_param():
    global param, offset_str
    if os.path.exists(cwd_path / "eFinder.config") == True:
        with open(cwd_path / "eFinder.config") as h:
            for line in h:
                line = line.strip("\n").split(":")
                param[line[0]] = str(line[1])
        pix_x, pix_y, dxstr, dystr = common.dxdy2pixel(
            float(param["d_x"]), float(param["d_y"])
        )
        offset_str = dxstr + "," + dystr


def save_param():
    global param
    with open(cwd_path / "eFinder.config", "w") as h:
        for key, value in param.items():
            # print("%s:%s\n" % (key, value))
            h.write("%s:%s\n" % (key, value))


def reader():
    global button
    while True:
        if handpad.get_box() in select.select([handpad.get_box()], [], [], 0)[0]:
            button = handpad.get_box().readline().decode("ascii").strip("\r\n")
        time.sleep(0.1)


# main code starts here

# time.sleep(1)
handpad = Display.Handpad(version)
coordinates = Coordinates.Coordinates()
nexus = Nexus.Nexus(handpad, coordinates)
nexus.read()
param = dict()
get_param()

handpad.display("ScopeDog eFinder", "Ready", "")
# array determines what is displayed, computed and what each button does for each screen.
# [first line,second line,third line, up button action,down...,left...,right...,select button short press action, long press action]
# empty string does nothing.
# example: left_right(-1) allows left button to scroll to the next left screen
# button texts are infact def functions
p = ""
home = [
    "ScopeDog",
    "eFinder",
    "ver" + version,
    "",
    "up_down(1)",
    "",
    "left_right(1)",
    "go_solve()",
    "",
]
nex = [
    "Nex: RA ",
    "    Dec ",
    "",
    "",
    "",
    "left_right(-1)",
    "left_right(1)",
    "go_solve()",
    "goto()",
]
sol = [
    "No solution yet",
    "'select' solves",
    "",
    "",
    "",
    "left_right(-1)",
    "left_right(1)",
    "go_solve()",
    "goto()",
]
delta = [
    "Delta: No solve",
    "'select' solves",
    "",
    "",
    "",
    "left_right(-1)",
    "left_right(1)",
    "go_solve()",
    "goto()",
]
aligns = [
    "'Select' aligns",
    "not aligned yet",
    str(p),
    "",
    "",
    "left_right(-1)",
    "left_right(1)",
    "align()",
    "",
]
polar = [
    "'Select' Polaris",
    offset_str,
    "",
    "",
    "",
    "left_right(-1)",
    "left_right(1)",
    "measure_offset()",
    "",
]
reset = [
    "'Select' Resets",
    offset_str,
    "",
    "",
    "",
    "left_right(-1)",
    "",
    "reset_offset()",
    "",
]
summary = ["", "", "", "up_down(-1)", "", "", "left_right(1)", "go_solve()", ""]
exp = [
    "Exposure",
    param["Exposure"],
    "",
    "up_down_inc(1,1)",
    "up_down_inc(1,-1)",
    "left_right(-1)",
    "left_right(1)",
    "go_solve()",
    "goto()",
]
gn = [
    "Gain",
    param["Gain"],
    "",
    "up_down_inc(2,1)",
    "up_down_inc(2,-1)",
    "left_right(-1)",
    "left_right(1)",
    "go_solve()",
    "goto()",
]
mode = [
    "Test mode",
    int(param["Test mode"]),
    "",
    "flip()",
    "flip()",
    "left_right(-1)",
    "left_right(1)",
    "go_solve()",
    "goto()",
]
status = [
    "Nexus via " + nexus.get_nexus_link(),
    "Nex align " + str(nexus.is_aligned()),
    "Brightness",
    "",
    "",
    "left_right(-1)",
    "",
    "go_solve()",
    "goto()",
]

arr = np.array(
    [
        [home, nex, sol, delta, aligns, polar, reset],
        [summary, exp, gn, mode, status, status, status],
    ]
)
update_summary()
deg_x, deg_y, dxstr, dystr = common.dxdy2pixel(float(param["d_x"]), float(param["d_y"]))
offset_str = dxstr + "," + dystr
new_arr = nexus.read_altAz(arr)
arr = new_arr
if nexus.is_aligned() == True:
    arr[0, 4][1] = "Nexus is aligned"
    arr[0, 4][0] = "'Select' syncs"


camera = common.pick_camera(param["Camera Type"], handpad, images_path)

handpad.display("ScopeDog eFinder", "v" + version, "")
button = ""
# main program loop, scan buttons and refresh display

scan = threading.Thread(target=reader)
scan.daemon = True
scan.start()

while True:  # next loop looks for button press and sets display option x,y
    if button == "21":
        exec(arr[x, y][7])
    elif button == "20":
        exec(arr[x, y][8])
    elif button == "19":
        exec(arr[x, y][4])
    elif button == "17":
        exec(arr[x, y][3])
    elif button == "16":
        exec(arr[x, y][5])
    elif button == "18":
        exec(arr[x, y][6])
    button = ""
    time.sleep(0.1)
